src/highlighters.cc

```diff
@@ -1088,28 +1105,55 @@ struct ShowWhitespacesHighlighter : Highlighter
                         if (it != end)
                             atom_it = line.split(atom_it, it.coord());
 
+                        const bool is_indent_guide = only_whitespace and whitespace_col % indentwidth == 0;
+
                         if (cp == '\t')
                         {
                             const ColumnCount column = get_column(buffer, tabstop, coord);
                             const ColumnCount count = tabstop - (column % tabstop) -
                                                       std::max(win_column - column, 0_col);
-                            atom_it->replace(m_tab + String(m_tabpad[(CharCount)0], count - m_tab.column_length()));
+                            if (is_indent_guide)
+                                atom_it->replace(m_guide + String(m_tabpad[(CharCount)0], count - m_guide.column_length()));
+                            else
+                                atom_it->replace(m_tab + String(m_tabpad[(CharCount)0], count - m_tab.column_length()));
+                            whitespace_col += count;
                         }
                         else if (cp == ' ')
-                            atom_it->replace(m_spc);
-                        else if (cp == '\n')
+                        {
+                            atom_it->replace(is_indent_guide ? m_guide : m_spc);
+                            whitespace_col++;
+                        }
+                        else
+                            only_whitespace = false;
+
+                        if (cp == '\n')
                             atom_it->replace(m_lf);
                         else if (cp == 0xA0 or cp == 0x202F)
                             atom_it->replace(m_nbsp);
-                        atom_it->face = merge_faces(atom_it->face, whitespaceface);
+
+                        atom_it->face = merge_faces(atom_it->face,
+                                                    only_whitespace and is_indent_guide ? indentguideface : whitespaceface);
                         break;
                     }
+                    else
+                        only_whitespace = false;
                 }
             }
         }
     }
 
-    const String m_tab, m_tabpad, m_spc, m_lf, m_nbsp;
+    bool all_whitespace(const StringView line, ColumnCount window_lhs)
+    {
+        for (auto it = line.begin(); it != line.end() and it - line.begin() <= window_lhs; ++it)
+        {
+            if (*it != '\t' and *it != ' ')
+                return false;
+        }
+        return true;
+    }
+
+    const String m_tab, m_tabpad, m_spc, m_lf, m_nbsp, m_guide;
+    const bool m_add_guide;
```

Comment:
I would prefer to use `m_guide.empty()` to check if we want to use the guide or not.